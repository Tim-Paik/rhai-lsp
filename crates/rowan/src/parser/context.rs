//! The parser context is a separate module to limit
//! the API surface for the parser functions.
#![allow(dead_code)]
use rowan::{GreenNodeBuilder, TextRange, TextSize};

use crate::syntax::{
    Lexer,
    SyntaxKind::{self, *},
};

use super::{Parse, ParseError, ParseErrorKind};

pub(crate) struct Context<'src> {
    lexer: Lexer<'src>,
    current_token: Option<SyntaxKind>,
    last_token: Option<SyntaxKind>,
    green: GreenNodeBuilder<'static>,
    errors: Vec<ParseError>,

    /// Tracks statements being separated by ";".
    statement_closed: bool,
}

impl<'src> Context<'src> {
    pub(crate) fn new(source: &'src str) -> Self {
        Self {
            lexer: Lexer::new(source),
            current_token: None,
            last_token: None,
            green: GreenNodeBuilder::new(),
            errors: Vec::new(),

            statement_closed: true,
        }
    }

    pub(crate) fn token(&mut self) -> Option<SyntaxKind> {
        // Eat insignificant tokens
        loop {
            if self.current_token.is_none() {
                self.current_token = self.lexer.next();
            }

            match self.current_token {
                Some(COMMENT_BLOCK) | Some(COMMENT_LINE) | Some(WHITESPACE) => {
                    self.eat();
                    self.last_token = None // We aren't interested in these.
                }
                Some(ERROR) => {
                    self.eat_error(ParseErrorKind::InvalidInput);
                    self.last_token = None // We aren't interested in these.
                }
                _ => break,
            }
        }

        self.current_token
    }

    pub(crate) fn previous_token(&self) -> Option<SyntaxKind> {
        self.last_token
    }

    pub(crate) fn eat(&mut self) {
        if let Some(t) = self.current_token.take() {
            self.green.token(t.into(), self.lexer.slice());
            self.last_token = Some(t);
        }
        self.current_token = None;
    }

    pub(crate) fn eat_as(&mut self, kind: SyntaxKind) {
        if self.current_token.is_some() {
            self.green.token(kind.into(), self.lexer.slice());
            self.last_token = Some(kind);
        }
        self.current_token = None;
    }

    pub(crate) fn eat_error(&mut self, error: ParseErrorKind) {
        self.add_error(error);
        self.eat();
    }

    pub(crate) fn eat_error_as(&mut self, kind: SyntaxKind, error: ParseErrorKind) {
        self.add_error(error);
        self.eat_as(kind);
    }

    pub(crate) fn add_error(&mut self, error: ParseErrorKind) {
        let span = self.lexer.span();
        self.errors.push(ParseError::new(
            TextRange::new(
                TextSize::from(span.start as u32),
                TextSize::from(span.end as u32),
            ),
            error,
        ));
    }

    pub(crate) fn start_node(&mut self, kind: SyntaxKind) {
        self.green.start_node(kind.into());
    }

    pub(crate) fn finish_node(&mut self) {
        self.green.finish_node();
    }

    pub(crate) fn finish(self) -> Parse {
        Parse {
            errors: self.errors,
            green: self.green.finish(),
        }
    }

    pub(crate) fn statement_closed(&self) -> bool {
        self.statement_closed
    }

    pub(crate) fn set_statement_closed(&mut self, v: bool) {
        self.statement_closed = v;
    }
}
