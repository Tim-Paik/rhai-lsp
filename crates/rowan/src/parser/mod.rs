use crate::syntax::{SyntaxKind, SyntaxNode};
use rowan::GreenNode;
use thiserror::Error;

use self::context::Context;

mod context;
pub mod parsers;

pub struct Parser<'src> {
    context: Context<'src>,
}

impl<'src> Parser<'src> {
    pub fn new(source: &'src str) -> Self {
        Self {
            context: Context::new(source),
        }
    }

    pub fn finish(self) -> Parse {
        self.context.finish()
    }
}

impl<'src> Parser<'src> {
    pub fn execute<F: FnOnce(&mut Context)>(&mut self, f: F) {
        f(&mut self.context)
    }
}

/// The result of parsing.
#[derive(Debug, Clone)]
pub struct Parse {
    pub errors: Vec<ParseError>,
    pub green: GreenNode,
}

impl Parse {
    pub fn into_syntax(self) -> SyntaxNode {
        SyntaxNode::new_root(self.green)
    }
}

#[derive(Debug, Clone, Error)]
#[error("parse error at {range:?}: {kind}")]
pub struct ParseError {
    pub range: rowan::TextRange,
    pub kind: ParseErrorKind,
}

impl ParseError {
    pub fn new(range: rowan::TextRange, kind: ParseErrorKind) -> Self {
        Self { range, kind }
    }
}

#[derive(Debug, Clone, Error)]
pub enum ParseErrorKind {
    #[error("unexpected EOF")]
    UnexpectedEof,

    #[error(r#"invalid input"#)]
    InvalidInput,

    #[error(r#"unexpected token"#)]
    UnexpectedToken,

    #[error(r#"expected token "{0:?}""#)]
    ExpectedToken(SyntaxKind),

    #[error(
        r#"expected one of "{}""#,
        .0
            .iter()
            .map(|s|format!("{:?}", s))
            .intersperse(String::from(", "))
            .collect::<String>()
    )]
    ExpectedOneOfTokens(Vec<SyntaxKind>),
}
