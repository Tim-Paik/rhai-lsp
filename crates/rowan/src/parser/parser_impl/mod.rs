#![deny(unreachable_patterns)]

use super::context::Context;
use crate::parser::ParseErrorKind;
use crate::syntax::SyntaxKind::*;
use crate::T;

macro_rules! require_token {
    ($ctx:ident) => {
        match $ctx.token() {
            Some(t) => t,
            None => return $ctx.eat_error(ParseErrorKind::UnexpectedEof),
        }
    };

    ($ctx:ident, finish_node) => {
        match $ctx.token() {
            Some(t) => t,
            None => {
                $ctx.finish_node();
                $ctx.eat_error(ParseErrorKind::UnexpectedEof);
                return;
            }
        }
    };
}

macro_rules! expect_token {
    ($ctx:ident, finish_node, $($token:tt)*) => {
        match $ctx.token() {
            Some(t) => match t {
                $($token)* => {
                    $ctx.eat();
                },
                _ => {
                    $ctx.finish_node();
                    $ctx.eat_error(ParseErrorKind::ExpectedToken($($token)*));
                    return;
                }
            }
            None => {
                $ctx.finish_node();
                $ctx.eat_error(ParseErrorKind::ExpectedToken($($token)*));
                return;
            }
        }
    };
    ($ctx:ident, $($token:tt)+) => {
        match $ctx.token() {
            Some(t) => match t {
                $($token)* => {
                    $ctx.eat();
                },
                _ => return $ctx.eat_error(ParseErrorKind::UnexpectedEof),
            }
            None => return $ctx.eat_error(ParseErrorKind::UnexpectedEof),
        }
    };

}

impl<'src> super::Parser<'src> {
    pub fn parse_file(&mut self) {
        parse_file(&mut self.context);
    }
}

fn parse_file(ctx: &mut Context) {
    ctx.start_node(FILE);
    if let Some(SHEBANG) = ctx.token() {
        parse_shebang(ctx);
    }

    ctx.set_statement_closed(true);
    while ctx.token().is_some() {
        if !ctx.statement_closed() {
            ctx.add_error(ParseErrorKind::ExpectedToken(T![";"]));
        }

        parse_stmt(ctx);
    }
    ctx.set_statement_closed(true);

    ctx.finish_node()
}

fn parse_shebang(ctx: &mut Context) {
    let token = require_token!(ctx);

    if !matches!(token, SHEBANG) {
        return ctx.eat_error(ParseErrorKind::ExpectedToken(SHEBANG));
    }

    ctx.eat()
}

fn parse_stmt(ctx: &mut Context) {
    let token = require_token!(ctx);
    ctx.start_node(STMT);
    ctx.set_statement_closed(false);

    if token == T![";"] {
        ctx.finish_node();
        ctx.set_statement_closed(true);
        return ctx.eat();
    }

    parse_item(ctx);

    if let Some(token) = ctx.token() {
        if token == T![";"] {
            ctx.eat();
            ctx.set_statement_closed(true);
        }
    }

    ctx.finish_node();
}

fn parse_item(ctx: &mut Context) {
    ctx.start_node(ITEM);

    // Parse doc comments if any.
    while matches!(
        require_token!(ctx, finish_node),
        COMMENT_BLOCK_DOC | COMMENT_LINE_DOC
    ) {
        ctx.eat();
    }

    parse_expr(ctx);

    ctx.finish_node();
}

fn parse_expr(ctx: &mut Context) {
    ctx.start_node(EXPR);
    let token = require_token!(ctx, finish_node);

    match token {
        T!["let"] => parse_expr_let(ctx),
        T!["const"] => parse_expr_const(ctx),
        T!["{"] => parse_expr_block(ctx),
        T!["fn"] => parse_expr_fn(ctx),
        LIT_INT | LIT_FLOAT | LIT_BOOL | LIT_STR | LIT_CHAR => parse_expr_lit(ctx),
        IDENT => parse_expr_ident(ctx),
        _ => ctx.eat_error(ParseErrorKind::UnexpectedToken),
    }

    ctx.finish_node();
}

fn parse_expr_ident(ctx: &mut Context) {
    ctx.start_node(EXPR_IDENT);
    parse_ident(ctx);
    ctx.finish_node();
}

fn parse_expr_lit(ctx: &mut Context) {
    ctx.start_node(EXPR_LIT);
    parse_lit(ctx);
    ctx.finish_node();
}

fn parse_expr_let(ctx: &mut Context) {
    ctx.start_node(EXPR_LET);

    expect_token!(ctx, finish_node, T!["let"]);
    expect_token!(ctx, finish_node, T!["ident"]);
    expect_token!(ctx, finish_node, T!["="]);

    parse_expr(ctx);

    ctx.finish_node();
}

fn parse_expr_const(ctx: &mut Context) {
    ctx.start_node(EXPR_CONST);

    expect_token!(ctx, finish_node, T!["const"]);
    expect_token!(ctx, finish_node, T!["ident"]);
    expect_token!(ctx, finish_node, T!["="]);

    parse_expr(ctx);

    ctx.finish_node();
}

fn parse_expr_block(ctx: &mut Context) {
    ctx.start_node(EXPR_BLOCK);

    expect_token!(ctx, finish_node, T!["{"]);

    ctx.set_statement_closed(true);
    loop {
        let token = require_token!(ctx, finish_node);

        if token == T!["}"] {
            break;
        }

        if !ctx.statement_closed() {
            ctx.add_error(ParseErrorKind::ExpectedToken(T![";"]));
        }

        parse_stmt(ctx);
    }

    expect_token!(ctx, finish_node, T!["}"]);

    // Blocks also don't require ";" when used as statements.
    ctx.set_statement_closed(true);

    ctx.finish_node();
}

fn parse_expr_fn(ctx: &mut Context) {
    ctx.start_node(EXPR_FN);

    expect_token!(ctx, finish_node, T!["fn"]);
    expect_token!(ctx, finish_node, T!["ident"]);

    parse_param_list(ctx);
    parse_expr_block(ctx);

    ctx.finish_node();
}

fn parse_param_list(ctx: &mut Context) {
    ctx.start_node(PARAM_LIST);

    let start_token = require_token!(ctx, finish_node);

    match start_token {
        T!["("] | T!["|"] => {
            ctx.eat();
        }
        _ => {
            ctx.eat_error(ParseErrorKind::ExpectedOneOfTokens(vec![T!["("], T!["|"]]));
            return;
        }
    }

    loop {
        let token = require_token!(ctx, finish_node);
        if matches!(
            (start_token, token),
            (T!["("], T![")"]) | (T!["|"], T!["|"])
        ) {
            ctx.eat();
            break;
        }

        parse_param(ctx);

        let end_token = require_token!(ctx, finish_node);

        match (start_token, end_token) {
            (T!["("], T![")"]) | (T!["|"], T!["|"]) => {
                ctx.eat();
                break;
            }
            (_, T![","]) => {
                ctx.eat();
            }
            _ => {
                ctx.eat_error(ParseErrorKind::ExpectedToken(T![","]));
                break;
            }
        }
    }

    ctx.finish_node();
}

fn parse_param(ctx: &mut Context) {
    ctx.start_node(PARAM);

    expect_token!(ctx, finish_node, T!["ident"]);

    ctx.finish_node();
}

fn parse_lit(ctx: &mut Context) {
    ctx.start_node(LIT);
    let token = require_token!(ctx, finish_node);

    match token {
        LIT_INT | LIT_FLOAT | LIT_BOOL | LIT_STR | LIT_CHAR => {
            ctx.eat();
        }
        _ => {
            ctx.eat_error(ParseErrorKind::ExpectedOneOfTokens(vec![
                LIT_INT, LIT_FLOAT, LIT_BOOL, LIT_STR, LIT_CHAR,
            ]));
        }
    }

    ctx.finish_node();
}

fn parse_ident(ctx: &mut Context) {
    let token = require_token!(ctx);

    match token {
        IDENT => ctx.eat(),
        _ => ctx.eat_error(ParseErrorKind::ExpectedToken(IDENT)),
    }
}

// Expr =
// | ExprUni
// | ExprBin
// | ExprParen
// | ExprArray
// | ExprIndex
// | ExprObject
// | ExprCall
// | ExprMethodCall
// | ExprAccess
// | ExprClosure
// | ExprIf
// | ExprLoop
// | ExprFor
// | ExprWhile
// | ExprBreak
// | ExprContinue
// | ExprSwitch
// | ExprReturn
// | ExprPath
// | ExprImport
