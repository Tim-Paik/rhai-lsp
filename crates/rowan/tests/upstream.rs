use rhai_rowan::parser::Parser;
use test_case::test_case;

#[test_case(include_str!("../../../rhai/scripts/array.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/assignment.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/comments.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/fibonacci.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/for1.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/for2.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/function_decl1.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/function_decl2.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/function_decl3.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/function_decl4.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/if1.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/if2.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/loop.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/mat_mul.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/module.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/oop.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/op1.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/op2.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/op3.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/primes.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/speed_test.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/string.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/strings_map.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/switch.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/while.rhai"))]
fn parse_rhai_upstream(src: &str) {
    let parse = Parser::new(src).parse_script();
    assert!(parse.errors.is_empty());
    // TODO(tests)
    // insta::assert_snapshot!(format!("{:#?}", parse.into_syntax()));
}
