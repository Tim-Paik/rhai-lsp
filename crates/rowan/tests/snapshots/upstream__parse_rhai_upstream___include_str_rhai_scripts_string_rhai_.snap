---
source: crates/rowan/tests/upstream.rs
expression: "format!(\"{:#?}\", parse.into_syntax())"

---
FILE@0..1384
  COMMENT_LINE@0..38 "// This script tests  ..."
  WHITESPACE@38..40 "\n\n"
  STMT@40..55
    ITEM@40..54
      EXPR@40..54
        EXPR_CALL@40..54
          EXPR@40..45
            EXPR_IDENT@40..45
              IDENT@40..45 "print"
          ARG_LIST@45..54
            PUNCT_PAREN_START@45..46 "("
            EXPR@46..53
              EXPR_LIT@46..53
                LIT@46..53
                  LIT_STR@46..53 "\"hello\""
            PUNCT_PAREN_END@53..54 ")"
    PUNCT_SEMI@54..55 ";"
  WHITESPACE@55..56 "\n"
  STMT@56..82
    ITEM@56..81
      EXPR@56..81
        EXPR_CALL@56..81
          EXPR@56..61
            EXPR_IDENT@56..61
              IDENT@56..61 "print"
          ARG_LIST@61..81
            PUNCT_PAREN_START@61..62 "("
            EXPR@62..80
              EXPR_LIT@62..80
                LIT@62..80
                  LIT_STR@62..80 "\"this\\nis \\\\ nice\""
            PUNCT_PAREN_END@80..81 ")"
    PUNCT_SEMI@81..82 ";"
  WHITESPACE@82..88 "      "
  COMMENT_LINE@88..107 "// escape sequences"
  WHITESPACE@107..108 "\n"
  STMT@108..134
    ITEM@108..133
      EXPR@108..133
        EXPR_CALL@108..133
          EXPR@108..113
            EXPR_IDENT@108..113
              IDENT@108..113 "print"
          ARG_LIST@113..133
            PUNCT_PAREN_START@113..114 "("
            EXPR@114..132
              EXPR_LIT@114..132
                LIT@114..132
                  LIT_STR@114..132 "\"0x40 hex is \\x40\""
            PUNCT_PAREN_END@132..133 ")"
    PUNCT_SEMI@133..134 ";"
  WHITESPACE@134..140 "      "
  COMMENT_LINE@140..162 "// hex escape sequence"
  WHITESPACE@162..163 "\n"
  STMT@163..192
    ITEM@163..191
      EXPR@163..191
        EXPR_CALL@163..191
          EXPR@163..168
            EXPR_IDENT@163..168
              IDENT@163..168 "print"
          ARG_LIST@168..191
            PUNCT_PAREN_START@168..169 "("
            EXPR@169..190
              EXPR_LIT@169..190
                LIT@169..190
                  LIT_STR@169..190 "\"Unicode fun: \\u2764\""
            PUNCT_PAREN_END@190..191 ")"
    PUNCT_SEMI@191..192 ";"
  WHITESPACE@192..195 "   "
  COMMENT_LINE@195..221 "// Unicode escape seq ..."
  WHITESPACE@221..222 "\n"
  STMT@222..252
    ITEM@222..251
      EXPR@222..251
        EXPR_CALL@222..251
          EXPR@222..227
            EXPR_IDENT@222..227
              IDENT@222..227 "print"
          ARG_LIST@227..251
            PUNCT_PAREN_START@227..228 "("
            EXPR@228..250
              EXPR_LIT@228..250
                LIT@228..250
                  LIT_STR@228..250 "\"more fun: \\U0001F603\""
            PUNCT_PAREN_END@250..251 ")"
    PUNCT_SEMI@251..252 ";"
  WHITESPACE@252..254 "  "
  COMMENT_LINE@254..280 "// Unicode escape seq ..."
  WHITESPACE@280..281 "\n"
  STMT@281..308
    ITEM@281..307
      EXPR@281..307
        EXPR_CALL@281..307
          EXPR@281..286
            EXPR_IDENT@281..286
              IDENT@281..286 "print"
          ARG_LIST@286..307
            PUNCT_PAREN_START@286..287 "("
            EXPR@287..306
              EXPR_BINARY@287..306
                EXPR_BINARY@287..299
                  EXPR_LIT@287..292
                    LIT@287..292
                      LIT_STR@287..292 "\"foo\""
                  WHITESPACE@292..293 " "
                  OP_ADD@293..294 "+"
                  EXPR@294..299
                    WHITESPACE@294..295 " "
                    EXPR_LIT@295..298
                      LIT@295..298
                        LIT_STR@295..298 "\" \""
                    WHITESPACE@298..299 " "
                OP_ADD@299..300 "+"
                EXPR@300..306
                  WHITESPACE@300..301 " "
                  EXPR_LIT@301..306
                    LIT@301..306
                      LIT_STR@301..306 "\"bar\""
            PUNCT_PAREN_END@306..307 ")"
    PUNCT_SEMI@307..308 ";"
  WHITESPACE@308..313 "     "
  COMMENT_LINE@313..345 "// string building us ..."
  WHITESPACE@345..346 "\n"
  STMT@346..367
    ITEM@346..366
      EXPR@346..366
        EXPR_CALL@346..366
          EXPR@346..351
            EXPR_IDENT@346..351
              IDENT@346..351 "print"
          ARG_LIST@351..366
            PUNCT_PAREN_START@351..352 "("
            EXPR@352..365
              EXPR_BINARY@352..365
                EXPR_LIT@352..357
                  LIT@352..357
                    LIT_STR@352..357 "\"foo\""
                WHITESPACE@357..358 " "
                OP_LT@358..359 "<"
                EXPR@359..365
                  WHITESPACE@359..360 " "
                  EXPR_LIT@360..365
                    LIT@360..365
                      LIT_STR@360..365 "\"bar\""
            PUNCT_PAREN_END@365..366 ")"
    PUNCT_SEMI@366..367 ";"
  WHITESPACE@367..378 "           "
  COMMENT_LINE@378..398 "// string comparison"
  WHITESPACE@398..399 "\n"
  STMT@399..421
    ITEM@399..420
      EXPR@399..420
        EXPR_CALL@399..420
          EXPR@399..404
            EXPR_IDENT@399..404
              IDENT@399..404 "print"
          ARG_LIST@404..420
            PUNCT_PAREN_START@404..405 "("
            EXPR@405..419
              EXPR_BINARY@405..419
                EXPR_LIT@405..410
                  LIT@405..410
                    LIT_STR@405..410 "\"foo\""
                WHITESPACE@410..411 " "
                OP_GT_EQ@411..413 ">="
                EXPR@413..419
                  WHITESPACE@413..414 " "
                  EXPR_LIT@414..419
                    LIT@414..419
                      LIT_STR@414..419 "\"bar\""
            PUNCT_PAREN_END@419..420 ")"
    PUNCT_SEMI@420..421 ";"
  WHITESPACE@421..431 "          "
  COMMENT_LINE@431..451 "// string comparison"
  WHITESPACE@451..452 "\n"
  STMT@452..481
    ITEM@452..480
      EXPR@452..480
        EXPR_CALL@452..480
          EXPR@452..457
            EXPR_IDENT@452..457
              IDENT@452..457 "print"
          ARG_LIST@457..480
            PUNCT_PAREN_START@457..458 "("
            EXPR@458..479
              EXPR_BINARY@458..479
                EXPR_LIT@458..474
                  LIT@458..474
                    LIT_STR@458..474 "\"the answer is \""
                WHITESPACE@474..475 " "
                OP_ADD@475..476 "+"
                EXPR@476..479
                  WHITESPACE@476..477 " "
                  EXPR_LIT@477..479
                    LIT@477..479
                      LIT_INT@477..479 "42"
            PUNCT_PAREN_END@479..480 ")"
    PUNCT_SEMI@480..481 ";"
  WHITESPACE@481..484 "   "
  COMMENT_LINE@484..525 "// string building us ..."
  WHITESPACE@525..527 "\n\n"
  STMT@527..569
    ITEM@527..568
      EXPR@527..568
        EXPR_LET@527..568
          KW_LET@527..530 "let"
          WHITESPACE@530..531 " "
          IDENT@531..532 "s"
          WHITESPACE@532..533 " "
          OP_ASSIGN@533..534 "="
          EXPR@534..568
            WHITESPACE@534..535 " "
            EXPR_LIT@535..568
              LIT@535..568
                LIT_STR@535..568 "\"\\u2764 hello, world! ..."
    PUNCT_SEMI@568..569 ";"
  WHITESPACE@569..570 " "
  COMMENT_LINE@570..588 "// string variable"
  WHITESPACE@588..589 "\n"
  STMT@589..614
    ITEM@589..613
      EXPR@589..613
        EXPR_CALL@589..613
          EXPR@589..594
            EXPR_IDENT@589..594
              IDENT@589..594 "print"
          ARG_LIST@594..613
            PUNCT_PAREN_START@594..595 "("
            EXPR@595..612
              EXPR_LIT@595..612
                LIT@595..612
                  LIT_STR@595..612 "`length=${s.len}`"
            PUNCT_PAREN_END@612..613 ")"
    PUNCT_SEMI@613..614 ";"
  WHITESPACE@614..621 "       "
  COMMENT_LINE@621..636 "// should be 17"
  WHITESPACE@636..638 "\n\n"
  STMT@638..655
    ITEM@638..654
      EXPR@638..654
        EXPR_BINARY@638..654
          EXPR_INDEX@638..648
            EXPR@638..639
              EXPR_IDENT@638..639
                IDENT@638..639 "s"
            PUNCT_BRACKET_START@639..640 "["
            EXPR@640..647
              EXPR_BINARY@640..647
                EXPR_BINARY@640..645
                  EXPR_IDENT@640..641
                    IDENT@640..641 "s"
                  PUNCT_DOT@641..642 "."
                  EXPR@642..645
                    EXPR_IDENT@642..645
                      IDENT@642..645 "len"
                OP_SUB@645..646 "-"
                EXPR@646..647
                  EXPR_LIT@646..647
                    LIT@646..647
                      LIT_INT@646..647 "3"
            PUNCT_BRACKET_END@647..648 "]"
          WHITESPACE@648..649 " "
          OP_ASSIGN@649..650 "="
          EXPR@650..654
            WHITESPACE@650..651 " "
            EXPR_LIT@651..654
              LIT@651..654
                LIT_CHAR@651..654 "'?'"
    PUNCT_SEMI@654..655 ";"
  WHITESPACE@655..670 "               "
  COMMENT_LINE@670..690 "// change the string"
  WHITESPACE@690..691 "\n"
  STMT@691..715
    ITEM@691..714
      EXPR@691..714
        EXPR_CALL@691..714
          EXPR@691..696
            EXPR_IDENT@691..696
              IDENT@691..696 "print"
          ARG_LIST@696..714
            PUNCT_PAREN_START@696..697 "("
            EXPR@697..713
              EXPR_LIT@697..713
                LIT@697..713
                  LIT_STR@697..713 "`Question: ${s}`"
            PUNCT_PAREN_END@713..714 ")"
    PUNCT_SEMI@714..715 ";"
  WHITESPACE@715..723 "        "
  COMMENT_LINE@723..764 "// should print 'Ques ..."
  WHITESPACE@764..766 "\n\n"
  COMMENT_LINE@766..787 "// Line continuation:"
  WHITESPACE@787..788 "\n"
  STMT@788..878
    ITEM@788..877
      EXPR@788..877
        EXPR_LET@788..877
          KW_LET@788..791 "let"
          WHITESPACE@791..792 " "
          IDENT@792..793 "s"
          WHITESPACE@793..794 " "
          OP_ASSIGN@794..795 "="
          EXPR@795..877
            WHITESPACE@795..796 " "
            EXPR_LIT@796..877
              LIT@796..877
                LIT_STR@796..877 "\"This is a long \\\n    ..."
    PUNCT_SEMI@877..878 ";"
  WHITESPACE@878..880 "\n\n"
  COMMENT_LINE@880..903 "// String interpolation"
  WHITESPACE@903..904 "\n"
  STMT@904..930
    ITEM@904..929
      EXPR@904..929
        EXPR_CALL@904..929
          EXPR@904..909
            EXPR_IDENT@904..909
              IDENT@904..909 "print"
          ARG_LIST@909..929
            PUNCT_PAREN_START@909..910 "("
            EXPR@910..928
              EXPR_LIT@910..928
                LIT@910..928
                  LIT_STR@910..928 "`One string: ${s}`"
            PUNCT_PAREN_END@928..929 ")"
    PUNCT_SEMI@929..930 ";"
  WHITESPACE@930..932 "\n\n"
  COMMENT_LINE@932..961 "// Multi-line literal ..."
  WHITESPACE@961..962 "\n"
  STMT@962..1108
    ITEM@962..1107
      EXPR@962..1107
        EXPR_LET@962..1107
          KW_LET@962..965 "let"
          WHITESPACE@965..966 " "
          IDENT@966..967 "s"
          WHITESPACE@967..968 " "
          OP_ASSIGN@968..969 "="
          EXPR@969..1107
            WHITESPACE@969..970 " "
            EXPR_LIT@970..1107
              LIT@970..1107
                LIT_STR@970..1107 "`\n          \\U0001F60 ..."
    PUNCT_SEMI@1107..1108 ";"
  WHITESPACE@1108..1110 "\n\n"
  STMT@1110..1119
    ITEM@1110..1118
      EXPR@1110..1118
        EXPR_CALL@1110..1118
          EXPR@1110..1115
            EXPR_IDENT@1110..1115
              IDENT@1110..1115 "print"
          ARG_LIST@1115..1118
            PUNCT_PAREN_START@1115..1116 "("
            EXPR@1116..1117
              EXPR_IDENT@1116..1117
                IDENT@1116..1117 "s"
            PUNCT_PAREN_END@1117..1118 ")"
    PUNCT_SEMI@1118..1119 ";"
  WHITESPACE@1119..1121 "\n\n"
  COMMENT_LINE@1121..1137 "// Interpolation"
  WHITESPACE@1137..1138 "\n"
  STMT@1138..1349
    ITEM@1138..1348
      EXPR@1138..1348
        EXPR_LET@1138..1348
          KW_LET@1138..1141 "let"
          WHITESPACE@1141..1142 " "
          IDENT@1142..1143 "s"
          WHITESPACE@1143..1144 " "
          OP_ASSIGN@1144..1145 "="
          EXPR@1145..1348
            WHITESPACE@1145..1146 " "
            EXPR_LIT@1146..1348
              LIT@1146..1348
                LIT_STR@1146..1348 "`This is interpolatio ..."
    PUNCT_SEMI@1348..1349 ";"
  WHITESPACE@1349..1351 "\n\n"
  STMT@1351..1360
    ITEM@1351..1359
      EXPR@1351..1359
        EXPR_CALL@1351..1359
          EXPR@1351..1356
            EXPR_IDENT@1351..1356
              IDENT@1351..1356 "print"
          ARG_LIST@1356..1359
            PUNCT_PAREN_START@1356..1357 "("
            EXPR@1357..1358
              EXPR_IDENT@1357..1358
                IDENT@1357..1358 "s"
            PUNCT_PAREN_END@1358..1359 ")"
    PUNCT_SEMI@1359..1360 ";"
  WHITESPACE@1360..1362 "\n\n"
  STMT@1362..1383
    ITEM@1362..1382
      EXPR@1362..1382
        EXPR_CALL@1362..1382
          EXPR@1362..1367
            EXPR_IDENT@1362..1367
              IDENT@1362..1367 "print"
          ARG_LIST@1367..1382
            PUNCT_PAREN_START@1367..1368 "("
            EXPR@1368..1381
              EXPR_LIT@1368..1381
                LIT@1368..1381
                  LIT_STR@1368..1381 "\">>> END <<<\""
            PUNCT_PAREN_END@1381..1382 ")"
    PUNCT_SEMI@1382..1383 ";"
  WHITESPACE@1383..1384 "\n"

